File: src/bmsx/render/runtime_asset_textures.ts
import { taskGate } from '../core/taskgate';
import type { AssetEntry, Memory } from '../machine/memory/memory';
import type { TextureManager } from './texture_manager';

const runtimeTextureEditGate = taskGate.group('asset:texture-update');

export function syncDirtyRuntimeImageAssets(memory: Memory, dirtyAssets: Iterable<AssetEntry>, texmanager: TextureManager): void {
	for (const entry of dirtyAssets) {
		if (entry.type !== 'image') {
			continue;
		}
		const vramSpan = entry.capacity > 0 ? entry.capacity : 1;
		if (memory.isVramRange(entry.baseAddr, vramSpan)) {
			continue;
		}
		const token = runtimeTextureEditGate.begin({ blocking: false, category: 'texture', tag: `asset:${entry.id}` });
		void texmanager.updateTexturesForKey(entry.id, memory.getImagePixels(entry), entry.regionW, entry.regionH)
			.finally(() => runtimeTextureEditGate.end(token));
	}
}

File: src/bmsx_cpp/render/runtime_asset_textures.cpp
#include "render/runtime_asset_textures.h"

#include "machine/memory/memory.h"
#include "render/texture_manager.h"

namespace bmsx {
namespace RuntimeAssetTextures {

void syncDirtyRuntimeImageAssets(Memory& memory, const std::vector<Memory::AssetEntry*>& dirtyAssets, TextureManager& texmanager) {
	for (const auto* entry : dirtyAssets) {
		if (entry->type != Memory::AssetType::Image) {
			continue;
		}
		const uint32_t span = entry->capacity > 0 ? entry->capacity : 1u;
		if (memory.isVramRange(entry->baseAddr, span)) {
			continue;
		}
		texmanager.updateTexturesForAsset(
			entry->id,
			memory.getImagePixels(*entry),
			static_cast<i32>(entry->regionW),
			static_cast<i32>(entry->regionH)
		);
	}
}

} // namespace RuntimeAssetTextures
} // namespace bmsx
